apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/ValueSerializer.cs

```diff
@@ -0,0 +1,150 @@
+    // TODO: Deliberate pass through of Value to Value (and maps/collections) to make it easier to plug in other mappers?
+
+    /// <summary>
+    /// Provides conversions from .NET types to Firestore Value protos.
+    /// </summary>
+    internal static class ValueSerializer
+    {
+        // Extension method to make the collection initializer below work
+        private static void Add<T>(this Dictionary<System.Type, Func<object, Value>> dictionary, Func<T, Value> converter) =>
+            dictionary.Add(typeof(T), obj => converter((T) obj));
+
+        private static readonly Dictionary<System.Type, Func<object, Value>> s_fromBclConverters =
+            new Dictionary<System.Type, Func<object, Value>>
+        {
+            (byte x) => new Value { IntegerValue = x },
+            (sbyte x) => new Value { IntegerValue = x },
+            (short x) => new Value { IntegerValue = x },
+            (ushort x) => new Value { IntegerValue = x },
+            (int x) => new Value { IntegerValue = x },
+            (uint x) => new Value { IntegerValue = x },
+            (long x) => new Value { IntegerValue = x },
+            (ulong x) => new Value { IntegerValue = checked ((long) x) },
+            (string x) => new Value { StringValue = x },
+            (float x) => new Value { DoubleValue = x },
+            (double x) => new Value { DoubleValue = x },
+            (bool x) => new Value { BooleanValue = x },
+            // TODO: Clone?
+            (wkt::Timestamp x) => new Value { TimestampValue = x },
+            (Timestamp x) => new Value { TimestampValue = x.ToProto() },
+            (DateTime x) => new Value { TimestampValue = wkt::Timestamp.FromDateTime(x) },
+            (DateTimeOffset x) => new Value { TimestampValue = wkt::Timestamp.FromDateTimeOffset(x) },
+            (byte[] x) => new Value { BytesValue = ByteString.CopyFrom(x) },
+            (ByteString x) => new Value { BytesValue = x },
+            (Blob x) => new Value { BytesValue = x.ByteString },
+            // TODO: Clone?
+            (LatLng x) => new Value { GeoPointValue = x },
+            (GeoPoint x) => new Value { GeoPointValue = x.ToProto() },
+            (DocumentReference x) => new Value { ReferenceValue = x.Path }
+        };
+
+        /// <summary>
+        /// Serializes a single input to a Value.
+        /// </summary>
+        /// <param name="value">The value to serialize.</param>
+        /// <returns>A Firestore Value proto.</returns>
+        internal static Value Serialize(object value)
+        {
+            if (value == null)
+            {
+                return new Value { NullValue = wkt::NullValue.NullValue };
+            }
+            if (s_fromBclConverters.TryGetValue(value.GetType(), out var converter))
+            {
+                return converter(value);
+            }
+            switch (value)
+            {
+                // TODO: What about other dictionary types (e.g Dictionary<string, string>)?
```

Comment:
Makes sense. I'd like to make that change in a follow-up PR if possible - this is all fiddly enough that it would be easy to mess up what's already working if I'm not careful.